src/native/libs/System.Globalization.Native/pal_casing.m

```diff
@@ -90,12 +106,28 @@ int32_t GlobalizationNative_ChangeCaseInvariantNative(const uint16_t* lpSrc, int
 
     int32_t srcIdx = 0, dstIdx = 0, isError = 0;
     uint16_t dstCodepoint;
-    while (srcIdx < result.length)
+    if (result.length <= cwDstLength)
+    {
+        while (srcIdx < result.length)
+        {
+            dstCodepoint = [result characterAtIndex:srcIdx++];
+            Append(lpDst, dstIdx, cwDstLength, dstCodepoint, isError);
+            if (isError)
+                return isError;
+        }
+    }
+    else
     {
-        dstCodepoint = [result characterAtIndex:srcIdx++];
-        Append(lpDst, dstIdx, cwDstLength, dstCodepoint, isError);
-        if (isError)
-            return isError;
+        while (srcIdx < cwSrcLength)
+        {           
+            NSString *src = [NSString stringWithCharacters: lpSrc + srcIdx length: 1];
+            srcIdx++;
+            NSString *dst = bToUpper ? src.uppercaseString : src.lowercaseString;
+            dstCodepoint = dst.length > 1 ? [src characterAtIndex: 0] : [dst characterAtIndex: 0];
+            Append(lpDst, dstIdx, cwDstLength, dstCodepoint, isError);
+            if (isError)
+                return isError;
+        }
```

Comment:
Is this is the same as [what is done in `pal_casing.c`](https://github.com/dotnet/runtime/blob/672ff9b9c666c5d0eb27cf2e6f3f8447e21004d0/src/native/libs/System.Globalization.Native/pal_casing.c#L55-L61) except we prevent any codepoint expansion?

I'm wondering if we actually need the above if block at all. `cwDstLength` is independent from the source and result strings. If we want to prevent any character code point expansions in general, the top block might permit character code point expansions in scenarios like: 1) `cwDstLength` is oversized  2) There happens to be characters that shrink in number of code points at the same time that other characters expand in the number of code points (not sure if that is possible). In either case, within each while loop we iterate through each character in the string, so the performance diff is `stringWithCharacters` + `.[upper/lower]caseString`. Not sure if those would be costly enough to warrant having the above `if` block, but I think because of 1) and 2) that we should remove it.

Reply:
ad. 1) why is `cwDstLength` oversized? Is it possible to maintain the behavior from ICU where we provide the function with a result buffer of exactly the size of the input string?
ad. 2) could you please provide an example of character that shrinks this way? Using e.g. `new CultureInfo("de-DE").TextInfo.ToLower("STRASSE")` should result in just `strasse` and I cannot find any proof that shrinking happens.

Reply:
Above if is used because this new added workaround is needed only for special cases, like string contains german `ß` or ligature characters like `ﬀ`. As these are special cases and in normal cases there is no need to take character one by one and uppercase/lowercase, above if will make uppercasing/lowercasing faster.
1) `cwDstLength` is not oversized. Here will still work same logic as in ICU, same way it will be passed to native function so result buffer has exactly the size of the input string.
2) I don't have examples that shrink that way.

Reply:
Updated function implementation similar to ICU, no more behavioural changes. Final sigma also works like ICU. Updated the doc.

Reply:
1. `cwDstLength` is passed in, so from the perspective of this function, it does not have to equal to `cwSrcLength` passed in. I meant that in the event that `cwDstLength` is larger (oversized) than `cwSrcLength`, the now removed if loop would have permitted code point expansions.
2. I was curious for a character that has a code point expansion when converting to upper/lower case, if the expanded version would "shrink" if we took the lower/upper case of that to get back to the original character. I was thinking this is the only other scenario in which `result.Length <= cwDstLength` (for a cwDstLength == cwSrcLength) would permit code point expansions (which it seemed we wanted to prohibit completely).